openmmtools/integrators.py

```diff
@@ -1895,6 +1905,145 @@ def _add_alchemical_reset_step(self):
+        """
+        # Check that a valid set of steps has been specified
+        nsteps_per_period = nsteps_eq + nsteps_neq + nsteps_eq + nsteps_neq
+        if nsteps_per_period <= 0:
+            raise ValueError(f'nsteps_per_period = {nsteps_per_period} must be > 0')
+
+        self._n_steps_eq = nsteps_eq # store number of equilibration steps to dwell within lambda = 0 or 1 when reached
+        self._n_steps_per_cycle = nsteps_eq + nsteps_neq + nsteps_eq + nsteps_neq
+        super().__init__(alchemical_functions, splitting, nsteps_neq=nsteps_neq, **kwargs)
+
+    def _add_global_variables(self):
+        """
+        modify the super (_add_global_variables) to add a collector for forward and reverse annealing, as well as to specify the number of equilibration
+        steps at endstates.
+        """
+        super()._add_global_variables()
+        self.addGlobalVariable('n_steps_eq', self._n_steps_eq)
+
+    def _add_integrator_steps(self):
+        """
+        Override the base class to insert reset steps around the integrator.
+        """
+        # First step: Constrain positions and velocities and reset work accumulators and alchemical integrators
+        self.beginIfBlock('step = 0')
+        self.addConstrainPositions()
+        self.addConstrainVelocities()
+        self._add_reset_protocol_work_step()
+        self._add_alchemical_reset_step()
+        self.endBlock()
+
+        # Main body
+        self.beginIfBlock("step >= 0")
+        super(AlchemicalNonequilibriumLangevinIntegrator, self)._add_integrator_steps()
+        self.addComputeGlobal("step", "step + 1")
+        self.endBlock()
+
+        # Reset step counter
+        self.beginIfBlock("step >= n_steps_per_cycle")
+        self.addComputeGlobal("step", "0")
+        self.addComputeGlobal("lambda_step", "0")
+        self.endBlock()
+
+        # Reset step
+        self.beginIfBlock('step = -1')
+        self._add_reset_protocol_work_step()
+        self._add_alchemical_reset_step() # sets step to 0
+        self.endBlock()
+    def _add_alchemical_perturbation_step(self):
+        """
+        Add alchemical perturbation step, accumulating protocol work.
+
+        .. todo ::
+
+           This could be extended to only evaluate energies for force groups with global context parameters
+           to speed up work evaluations.
+
+        """
+        # Store initial potential energy
+        self.addComputeGlobal("Eold", "energy")
+
+        # Update lambda and increment that tracks updates.
+        lambda_control_expression = "lambda"\
+            " + delta_lambda*step((step+0.5)-n_steps_eq)*step((n_steps_eq+n_steps_neq) - (step+0.5))"\
+            " - delta_lambda*step((step+0.5)-(n_steps_eq+n_steps_neq+n_steps_eq))*step(n_steps_per_cycle - (step+0.5));"\
+            "delta_lambda = 1/n_lambda_steps;" # Note that there may be multiple H per integrator string, so we correct for this here
```

Comment:
could you split this expression into substrings and provide a short comment on each part?  it's not too easy to read in this format

Reply:
Great idea!